View compilation fails when cross-compiling for Linux on Windows

1. Use Visual Studio to create a new ASP.NET Core Web Application (.NET Framework). Note: **Not** .NET Core, as I'm running my site on Mono with the full .NET Framework.
2. Add NuGet reference to `Microsoft.AspNetCore.Mvc.Razor.ViewCompilation` version 1.1.0
3. Modify the `csproj` to add:
```
<MvcRazorCompileOnPublish>true</MvcRazorCompileOnPublish>
```
4. Run
```
dotnet restore -r debian-x64
dotnet publish . -o "C:\temp\ViewPrecompilationBug_Deploy" -c Release -r debian-x64
```

Expected: Publish works as expected
Actual: Error is thrown:
```
C:\temp\ViewPrecompilationBug\ViewPrecompilationBug
λ dotnet publish . -o "C:\temp\ViewPrecompilationBug_Deploy" -c Release -r debian-x64
Microsoft (R) Build Engine version 15.1.548.43366
Copyright (C) Microsoft Corporation. All rights reserved.

  ViewPrecompilationBug -> C:\temp\ViewPrecompilationBug\ViewPrecompilationBug\bin\Release\net452\debian-x64\ViewPrecompilationBug.exe
  'bin\Release\net452\debian-x64\Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.exe' is not recognized as an internal or external command,
  operable program or batch file.
C:\Users\Daniel\.nuget\packages\microsoft.aspnetcore.mvc.razor.viewcompilation\1.1.0\build\net451\Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.targets(36,5): error MSB3073: The command "bin\Release\net452\debian-x64\Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.exe @"obj\Release\net452\debian-x64\microsoft.aspnetcore.mvc.razor.viewcompilation.rsp"" exited with code 9009. [C:\temp\ViewPrecompilationBug\ViewPrecompilationBug\ViewPrecompilationBug.csproj]
```

To expand a bit on what @simsod said. We are building this on OSX to be deployed on a Windows server.

``` dotnet publish -r osx.10.11-x64``` gives the following output:

```
Microsoft (R) Build Engine version 15.1.548.43366
Copyright (C) Microsoft Corporation. All rights reserved.

  Saga.Web.Core -> /Users/karl/src/Saga/Saga.Web/src/Saga.Web.Core/bin/Debug/netstandard1.6/Saga.Web.Core.dll

  Saga.Web -> /Users/karl/src/Saga/Saga.Web/src/Saga.Web/bin/Debug/netcoreapp1.1/osx.10.11-x64/Saga.Web.dll
  Razor view compilation for Saga.Web -> obj/Debug/netcoreapp1.1/osx.10.11-x64/Saga.Web.PrecompiledViews.dll
```

``` dotnet publish -r win7-x64``` on the other hand crahes like this:

```
Microsoft (R) Build Engine version 15.1.548.43366
Copyright (C) Microsoft Corporation. All rights reserved.

  Saga.Web.Core -> /Users/karl/src/Saga/Saga.Web/src/Saga.Web.Core/bin/Debug/netstandard1.6/Saga.Web.Core.dll

  Saga.Web -> /Users/karl/src/Saga/Saga.Web/src/Saga.Web/bin/Debug/netcoreapp1.1/win7-x64/Saga.Web.dll
  Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
/Users/karl/.nuget/packages/microsoft.aspnetcore.mvc.razor.viewcompilation/1.1.0/build/netcoreapp1.1/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.targets(25,5): error MSB3073: The command "dotnet exec --runtimeconfig "/Users/karl/src/Saga/Saga.Web/src/Saga.Web/bin/Debug/netcoreapp1.1/win7-x64/Saga.Web.runtimeconfig.json" --depsfile "/Users/karl/src/Saga/Saga.Web/src/Saga.Web/bin/Debug/netcoreapp1.1/win7-x64/Saga.Web.deps.json" "/Users/karl/.nuget/packages/microsoft.aspnetcore.mvc.razor.viewcompilation/1.1.0/build/netcoreapp1.1/../../lib/netcoreapp1.1/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.dll" @"obj/Debug/netcoreapp1.1/win7-x64/microsoft.aspnetcore.mvc.razor.viewcompilation.rsp"" exited with code 135. [/Users/karl/src/Saga/Saga.Web/src/Saga.Web/Saga.Web.csproj]
```

If test it out on a Windows 10 machine the opposite things work (ie win7-x64 compiles fine and osx.10.11-x64 fails with a similar error). If I disable viewcompilation everything is fine on both environments.

This worked fine using the old tooling (project.json) using the same version of ```Microsoft.AspNetCore.Mvc.Razor.ViewCompilation``` (ie 1.1.0).

--- a/src/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation/build/netstandard2.0/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.targets
+++ b/src/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation/build/netstandard2.0/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.targets
@@ -1,9 +1,8 @@
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Target Name="_ResolveInputArguments">
     <PropertyGroup>
-      <_MvcRazorOutputPath Condition="'$(MvcRazorOutputPath)'!=''">$([MSBuild]::EnsureTrailingSlash('$(MvcRazorOutputPath)'))</_MvcRazorOutputPath>
-      <_MvcRazorOutputPath Condition="'$(_MvcRazorOutputPath)'==''">$(IntermediateOutputPath)</_MvcRazorOutputPath>
-      <_MvcRazorOutputFullPath>$(_MvcRazorOutputPath)$(AssemblyName).PrecompiledViews.dll</_MvcRazorOutputFullPath>
+      <MvcRazorOutputPath Condition="'$(MvcRazorOutputPath)'==''">$(IntermediateOutputPath)</MvcRazorOutputPath>
+      <_MvcRazorOutputFullPath Condition="'$(_MvcRazorOutputFullPath)'==''">$([MSBuild]::EnsureTrailingSlash('$(MvcRazorOutputPath)'))$(AssemblyName).PrecompiledViews.dll</_MvcRazorOutputFullPath>
       <_MvcRazorResponseFilePath>$(IntermediateOutputPath)microsoft.aspnetcore.mvc.razor.viewcompilation.rsp</_MvcRazorResponseFilePath>
 
       <MvcRazorContentRoot Condition="'$(MvcRazorContentRoot)'==''">$(MSBuildProjectDirectory)</MvcRazorContentRoot>
@@ -25,7 +24,7 @@
     <CallTarget Targets="_MvcRazorPrecompile" />
   </Target>
 
-  <Target Name="_MvcRazorPrecompile" DependsOnTargets="_RunForCore;_RunForDesktop" />
+  <Target Name="_MvcRazorPrecompile" DependsOnTargets="_RunForCore;_RunForCoreWithRID;_RunForDesktop" />
 
   <PropertyGroup>
     <_MvcViewCompilationTasksPath Condition="'$(_MvcViewCompilationTasksPath)'==''">$(MSBuildThisFileDirectory)$(MSBuildThisFileName).Tasks.dll</_MvcViewCompilationTasksPath>
@@ -58,6 +57,22 @@
     </PropertyGroup>
 
     <Exec Command="$(ExecArgs)" WorkingDirectory="$(MSBuildProjectDirectory)" />
+  </Target>
+
+  <Target
+    Name="_RunForCoreWithRID"
+    DependsOnTargets="_ResolveInputArguments;_CreateResponseFileForMvcRazorPrecompile"
+    Condition="'$(TargetFrameworkIdentifier)'=='.NETCoreApp' AND '$(RuntimeIdentifier)'!=''">
+
+    <ItemGroup>
+      <_RazorCompilationProject Include="$(MSBuildProjectFullPath)">
+        <AdditionalProperties>RuntimeIdentifier=;MvcRazorOutputPath=$(MvcRazorOutputPath)</AdditionalProperties>
+      </_RazorCompilationProject>
+    </ItemGroup>
+
+    <MSBuild
+      Projects="@(_RazorCompilationProject)"
+      Targets="Build;MvcRazorPrecompile" />
   </Target>
 
   <Target
@@ -109,7 +124,7 @@
     <ItemGroup>
       <_ResponseFileLines Include="
         $(MSBuildProjectDirectory);
-        --output-path=$(_MvcRazorOutputPath);
+        --output-path=$(MvcRazorOutputPath);
         --application-name=$(AssemblyName);
         --content-root=$(MvcRazorContentRoot);" />
 
@@ -144,7 +159,7 @@
 
   <Target Name="_MvcRazorResolveFilesToCompute"
     AfterTargets="ComputeRefAssembliesToPublish"
-    Condition="'$(MvcRazorCompileOnPublish)'=='true' AND ('$(TargetFrameworkIdentifier)'!='.NETCoreApp' OR '$(RuntimeIdentifier)'=='')">
+    Condition="'$(MvcRazorCompileOnPublish)'=='true'">
 
     <PropertyGroup>
       <_MvcRazorOutputPdbFullPath>$([System.IO.Path]::ChangeExtension('$(_MvcRazorOutputFullPath)', '.pdb'))</_MvcRazorOutputPdbFullPath>

--- a/test/FunctionalTests/Infrastructure/ApplicationTestFixture.cs
+++ b/test/FunctionalTests/Infrastructure/ApplicationTestFixture.cs
@@ -29,18 +29,21 @@ namespace FunctionalTests
         protected abstract DeploymentParameters GetDeploymentParameters();
 
         protected DeploymentParameters GetDeploymentParameters(RuntimeFlavor flavor)
+            => GetDeploymentParameters(ApplicationPath, ApplicationName, flavor);
+
+        public static DeploymentParameters GetDeploymentParameters(string applicationPath, string applicationName, RuntimeFlavor flavor)
         {
             var telemetryOptOut = new KeyValuePair<string, string>(
                 DotnetCLITelemetryOptOut,
                 "1");
 
             var deploymentParameters = new DeploymentParameters(
-                ApplicationPath,
+                applicationPath,
                 ServerType.Kestrel,
                 flavor,
                 RuntimeArchitecture.x64)
             {
-                ApplicationName = ApplicationName,
+                ApplicationName = applicationName,
                 PublishApplicationBeforeDeployment = true,
                 TargetFramework = flavor == RuntimeFlavor.Clr ? "net461" : "netcoreapp2.0",
                 Configuration = "Release",

--- a/test/FunctionalTests/PublishForRidTest_CoreCLR.cs
+++ b/test/FunctionalTests/PublishForRidTest_CoreCLR.cs
@@ -1,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Testing;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace FunctionalTests
+{
+    public class PublishWithRidTest_CoreCLR : LoggedTest
+    {
+        public PublishWithRidTest_CoreCLR(ITestOutputHelper output)
+            : base(output)
+        {
+        }
+
+        [Fact]
+        public void CrossPublishingWorks()
+        {
+            using (StartLog(out var loggerFactory))
+            {
+                // Arrange
+                var applicationName = nameof(SimpleApp);
+                var applicationPath = ApplicationPaths.GetTestAppDirectory(applicationName);
+                var deploymentParameters = ApplicationTestFixture.GetDeploymentParameters(
+                    applicationPath, 
+                    applicationName, 
+                    RuntimeFlavor.CoreClr);
+                
+                // Deploy for a rid that does not exist on the current platform.
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    deploymentParameters.AdditionalPublishParameters = "-r debian-x64";
+                }
+                else
+                {
+                    deploymentParameters.AdditionalPublishParameters = "-r win7-x86";
+                }
+                var deployer = new DotNetPublishDeployer(deploymentParameters, loggerFactory);
+
+                // Act
+                deployer.DotnetPublish();
+
+                // Act
+                var expectedFile = Path.Combine(
+                    deploymentParameters.PublishedApplicationRootPath, 
+                    $"{applicationName}.PrecompiledViews.dll");
+                Assert.True(File.Exists(expectedFile), $"Expected precompiled file {expectedFile} does not exist.");
+            }
+        }
+
+        private class DotNetPublishDeployer : ApplicationDeployer
+        {
+            public DotNetPublishDeployer(DeploymentParameters deploymentParameters, ILoggerFactory loggerFactory) 
+                : base(deploymentParameters, loggerFactory)
+            {
+            }
+
+            public void DotnetPublish() => base.DotnetPublish();
+
+            public override void Dispose() => CleanPublishedOutput();
+
+            public override Task<DeploymentResult> DeployAsync() => throw new NotSupportedException();
+        }
+    }
+}

